src/parser.rs

```diff
@@ -0,0 +1,298 @@
+use crate::{
+    error::{error, ParserError},
+    grammar::{declaration, expression, statement},
+    token::{Token, TokenType},
+};
+
+pub struct Parser {
+    tokens: Vec<Token>,
+    current_pos: usize,
+}
+
+impl Parser {
+    pub fn new(tokens: Vec<Token>) -> Self {
+        Self {
+            tokens,
+            current_pos: 0,
+        }
+    }
+
+    fn error(&self, message: &str) {
+        if self.done() {
+            error(0, message)
+        }
+        error(self.get_current().line, message);
+    }
+
+    fn advance(&mut self) {
+        self.current_pos += 1;
+    }
+
+    fn get_previous(&self) -> Token {
+        (*self.tokens.get(self.current_pos - 1).unwrap()).clone()
+    }
+
+    fn get_current(&self) -> Token {
+        (*self.tokens.get(self.current_pos).unwrap()).clone()
+    }
+
+    fn done(&self) -> bool {
+        self.current_pos >= self.tokens.len()
+    }
+
+    fn consume_if<F>(&mut self, condition: F) -> bool
+    where
+        F: Fn(Token) -> bool,
+    {
+        if self.done() || !condition(self.get_current()) {
+            return false;
+        }
+
+        self.advance();
+        true
+    }
+
+    fn consume_if_is_type(&mut self, types: Vec<TokenType>) -> bool {
+        return self.consume_if(|t| types.iter().any(|x| *x == t.token_type));
```

Comment:
```suggestion
        self.consume_if(|t| types.iter().any(|x| *x == t.token_type))
```
don't need a `return`